### PR TITLE
Adds v11 schema

### PR DESCRIPTION
**Pull request by cyriltovena**
Signed-off-by: Cyril Tovena <cyril.tovena@gmail.com>

We ran Loki for a while using the new `LabelNamesForMetricName` to retrieve all possibles labels. Turned out that it does pull too many chunks for a single query spanning over 6h (12k~)

This introduces a new schema where we store the metric (labels set) within the `IndexEntry.Value`. Other schema keep the same implementation.

Design doc & reason for this change: https://docs.google.com/document/d/1sVZHtQACLQZfiKnnFhXSiqyRx6YSX945s9jPjc2js6o/edit?usp=sharing

/cc @gouthamve

### CODE REVIEW

**Comment by weeco** [comment]
@cyriltovena Can you make the design doc accessible for the public?

**Comment by cyriltovena** [comment]
Oups I though it was already sorry !

File: pkg/chunk/schema_test.go [hunk 1]
```diff
@@ -333,3 +335,55 @@ func TestSchemaRangeKey(t *testing.T) {
 		})
 	}
 }
+
+func BenchmarkEncodeLabelsJson(b *testing.B) {
```
Comment (cyriltovena) [inline pkg/chunk/schema_test.go]: This was a benchmark I used to decide if I should use json or string encoding for the labels set.

`.String()`/`promql.ParseMetric` is faster than `json.Marshal`/`json.Unmarshal`.

Comment (bboreham) [reply pkg/chunk/schema_test.go]: In case you didn't notice, `pkg/chunk` has special code to speed up the latter via `github.com/json-iterator/go`

Comment (cyriltovena) [reply pkg/chunk/schema_test.go]: I'll give it a look, thanks @bboreham .

**Comment by cyriltovena** [comment]
This would also help advance #416 !

File: pkg/chunk/schema_test.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/chunk/schema_test.go]: Wow 
```
BenchmarkEncodeLabelsString-16    	    1000	     13343 ns/op	    2029 B/op	      35 allocs/op
BenchmarkEncodeLabelsJson-16    	    1000	      1074 ns/op	     586 B/op	      18 allocs/op
```

@gouthamve I see you already approved I'm going to change to use json-iterator as suggested by @bboreham

Comment (cyriltovena) [reply pkg/chunk/schema_test.go]: done !

File: pkg/chunk/schema.go [hunk 2]
```diff
@@ -734,3 +774,47 @@ func (v10Entries) GetChunksForSeries(bucket Bucket, seriesID []byte) ([]IndexQue
...
+	}
+
+	return entries, nil
+}
+
+func (v11Entries) GetMetricsForSeries(bucket Bucket, seriesID []byte) ([]IndexQuery, error) {
+	return []IndexQuery{
+		{
+			TableName: bucket.tableName,
+			HashValue: bucket.hashKey + ":" + string(seriesID) + ":metric",
```
Comment (tomwilkie) [inline pkg/chunk/schema.go]: Do you need to specify a range value here?

File: pkg/distributor/ha_tracker.go [hunk 3]
```diff
@@ -172,6 +174,10 @@ func (c *haTracker) checkKVStore(ctx context.Context, key, replica string, now t
 	})
 }
 
+func replicasNotMatchError(replica, elected string) error {
+	return httpgrpc.Errorf(http.StatusAccepted, "replicas did not mach, rejecting sample: replica=%s, elected=%s", replica, elected)
+}
```
Comment (tomwilkie) [inline pkg/distributor/ha_tracker.go]: Spurious change?

**Comment by tomwilkie** [comment]
Is this covered by any end-to-end test?

**Comment by cyriltovena** [comment]
It is covered by unit tests, could you point me at the end-to-end test ?

File: pkg/chunk/schema.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/chunk/schema.go]: I added a `+ ":metric"` at the end of the HasValue, so I don't think I need the RangeValue to query. (From the doc above it means I'm getting all of them)

File: pkg/distributor/ha_tracker.go (hunk 3 above)
Comment (cyriltovena) [reply pkg/distributor/ha_tracker.go]: This appeared after I rebased amend a commit, I wanted to trick the CI to trigger a build as it was failing on a flaky test on the ruler.

So this is from master

File: pkg/chunk/schema.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/chunk/schema.go]: However the `IndexEntry` does have a RangeValue as it seems required for validation. But it is just a constant.

File: pkg/distributor/ha_tracker.go (hunk 3 above)
Comment (cyriltovena) [reply pkg/distributor/ha_tracker.go]: I rebased it again and its gone !

File: pkg/chunk/schema.go [hunk 4]
```diff
@@ -734,3 +774,47 @@ func (v10Entries) GetChunksForSeries(bucket Bucket, seriesID []byte) ([]IndexQue
...
+	}
+	entries := []IndexEntry{
+		// Entry for seriesID -> chunkID
+		{
+			TableName:  bucket.tableName,
+			HashValue:  bucket.hashKey + ":" + string(seriesID),
+			RangeValue: encodeRangeKey(encodedThroughBytes, nil, []byte(chunkID), chunkTimeRangeKeyV3),
+		},
+		// Entry for seriesID -> labels
+		{
```
Comment (gouthamve) [inline pkg/chunk/schema.go]: We should move this into `GetLabelWriteEntries` as those will be cached and deduped. I don't see a reason for this to be in `GetChunkWriteEntries`.

**Comment by cyriltovena** [comment]
@gouthamve I've added some improvements:

- The index entry is now only returned via GetLabelWriteEntries which means the entry is created only for a unique label set.
- I'm storing now only label names (not values) and skipping __name__ as it will always be there.
- deduping using a map.

File: pkg/chunk/schema.go (hunk 4 above)
Comment (cyriltovena) [reply pkg/chunk/schema.go]: done

**Comment by gouthamve** [comment]
Rebased and calculated the overhead. This adds 7% to the index size: https://docs.google.com/spreadsheets/d/1admjIY_SyBN8q7FJdhumNtsM85yOWxTgfCUinRtYLwY/edit

Calculated using v10 as base: https://github.com/cortexproject/cortex/blob/98db0a4e4f007c2dbce35efae23bb37b5ee650db/pkg/chunk/schema.go#L642-L673

**Comment by bboreham** [comment]
TIL that Google Sheets doesn't colour in the cells in a formula when you don't have edit permission.
So it's a bit hard to follow.

Came here to say that DynamoDB charges 100 bytes overhead per record, so the storage cost in both cases is quite a bit higher than the raw byte count, and the relative cost of 130 extra bytes in one record is much less.

**Comment by gouthamve** [comment]
I've made the sheet editable, but given that it's even better for Dynamo, can this get an LGTM? I wanna first roll this out and then experiment with a v12 that doesn't require a `__name__`.

**Comment by bboreham** [comment]
Should we have something in the docs saying which schemas are solid and which are experimental / not recommended ?

File: pkg/chunk/chunk_store_utils.go [hunk 5]
```diff
@@ -42,11 +42,9 @@ func labelNamesFromChunks(chunks []Chunk) []string {
 	var result []string
 	for _, c := range chunks {
 		for _, l := range c.Metric {
-			if l.Name != model.MetricNameLabel {
-				if _, ok := keys[string(l.Name)]; !ok {
-					keys[string(l.Name)] = struct{}{}
-					result = append(result, string(l.Name))
-				}
+			if _, ok := keys[string(l.Name)]; !ok {
```
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/chunk/chunk_store_utils.go]: This is a separate change yes, this is to maintain some compatibility with the Prometheus API which returns `__name__`

File: pkg/chunk/schema_config.go [hunk 6]
```diff
@@ -219,8 +219,12 @@ func (cfg *SchemaConfig) ForEachAfter(t model.Time, f func(config *PeriodConfig)
 
 // CreateSchema returns the schema defined by the PeriodConfig
 func (cfg PeriodConfig) CreateSchema() Schema {
-	var e entries
+	rowShards := uint32(16)
```
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/chunk/schema_config.go]: Hrm, so we load the schema from YAML. Now, we are defaulting it to `16`, but only here. On the other hand, we could loop through all the configs we loaded and set it to `16` after Unmarshaling, but I am not sure that is more cleaner than this.

**Comment by gouthamve** [comment]
> Also, am I right in thinking the labels data is never cached?

No, it should be cached as far as I can see. Atleast the index lookups will be cached and that should be enough, no?

**Comment by cyriltovena** [comment]
> > Also, am I right in thinking the labels data is never cached?
> 
> No, it should be cached as far as I can see. Atleast the index lookups will be cached and that should be enough, no?

I think index cache is enough, since we store the data in the index only.

File: pkg/chunk/chunk_store_utils.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bboreham) [reply pkg/chunk/chunk_store_utils.go]: Needs to be in the description and the changelog, no?

**Comment by bboreham** [comment]
OK, I think I understand how the caching works now.
I would still like to see the blind alleys removed from the commits.

File: pkg/chunk/chunk_store_utils.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/chunk/chunk_store_utils.go]: There's no user facing changes here. The API is not being used by Cortex yet and I don't see the reason to update the changelog.
